File: java-suite/src/main/java/de/fraunhofer/iem/reflection/ClassReference.java
package de.fraunhofer.iem.reflection;

/**
 * Demonstrates the Class reference in Java reflection
 */
public class ClassReference {
    /**
     * This method demonstrates the class reference directly using the class. Demonstrates both KClass and Java class reference.
     */
    public static void directClassAccess() {
        Class<ClassReference> clz = ClassReference.class;
        System.out.println(clz.getName() + " : " + clz.getClass().getName());
    }

    /**
     * This method demonstrates the class reference from the variable. Demonstrates both KClass and Java class reference.
     */
    public static void classAccessThroughVariable() {
        ClassReference classReference = new ClassReference();

        Class<? extends ClassReference> clz = classReference.getClass();
        System.out.println(clz.getName() + " : " + clz.getClass().getName());
    }

    public static void main(String[] args) {
        directClassAccess();
        classAccessThroughVariable();
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/reflection/ClassReference.kt
package de.fraunhofer.iem.reflection

/**
 * Dummy Class
 *
 * @author Ranjith Krishnamurthy
 */
class ClassReference {
}

fun main() {
    directClassAccess()
    classAccessThroughVariable()
}

/**
 * This method demonstrates the class reference directly using the class. Demonstrates both KClass and Java class reference.
 */
fun directClassAccess() {
    val kotlinClz = ClassReference::class
    val javaClz = ClassReference::class.java

    println(kotlinClz.qualifiedName + " : " + kotlinClz::class.qualifiedName)
    println(javaClz.name + " : " + javaClz::class.qualifiedName)
}

/**
 * This method demonstrates the class reference from the variable. Demonstrates both KClass and Java class reference.
 */
fun classAccessThroughVariable() {
    val ob1 = ClassReference()
    var ob2 = ClassReference()

    val kotlinClz1 = ob1::class
    val javaClz1 = ob1::class.java
    val kotlinClz2 = ob2::class
    val javaClz2 = ob2::class.java

    println(kotlinClz1.qualifiedName + " : " + kotlinClz1::class.qualifiedName)
    println(javaClz1.name + " : " + javaClz1::class.qualifiedName)
    println(kotlinClz2.qualifiedName + " : " + kotlinClz2::class.qualifiedName)
    println(javaClz2.name + " : " + javaClz2::class.qualifiedName)
}